doc/man/man8/SGE_Helper_Service.exe.8
'\" t
.\"___INFO__MARK_BEGIN__
.\"
.\" Copyright: 2004 by Sun Microsystems, Inc.
.\"
.\" Portions of this software are Copyright (c) 2011 Univa Corporation
.\"
.\"___INFO__MARK_END__
.\" $RCSfile: SGE_Helper_Service.exe.8,v $     Last Update: $Date: 2011-06-22 17:57:39 $     Revision: $Revision: 1.4 $
.\"
.\"
.\" Some handy macro definitions [from Tom Christensen's man(1) manual page].
.\"
.de SB		\" small and bold
.if !"\\$1"" \\s-2\\fB\&\\$1\\s0\\fR\\$2 \\$3 \\$4 \\$5
..
.\"
.de T		\" switch to typewriter font
.ft CW		\" probably want CW if you don't have TA font
..
.\" "
.de TY		\" put $1 in typewriter font
.if t .T
.if n ``\c
\\$1\c
.if t .ft P
.if n \&''\c
\\$2
..
.\"
.de M		\" man page reference
\\fI\\$1\\fR\\|(\\$2)\\$3
..
.TH SGE_Helper_Service.exe 8 "$Date: 2011-06-22 17:57:39 $" "xxRELxx" "xxQS_NAMExx Administrative Commands"
.SH NAME
SGE_Helper_Service.exe \- xxQS_NAMExx job execution agent helper service
.\"
.\"
.SH SYNOPSIS
.B SGE_Helper_Service.exe
[
.B \-help
|
.B \-install 
|
.B \-uninstall
]
.\"
.\"
.SH DESCRIPTION
.PP
.I SGE_Helper_Service.exe helps the 
.M xxqs_name_sxx_execd 8
to run Windows jobs that request to be allowed to display
their GUI on the visible Desktop. It is a Win32 service that is started at boot time
under the SYSTEM account, giving it the permissions to give any process the 
permissions to access the visible Desktop.
.br
This service runs independently of the
.M xxqs_name_sxx_execd 8
(i.e. it is not started and stopped together with the execution daemon) and there
can be only one
.I SGE_Helper_Service.exe
per host.
.PP
Jobs can request to be allowed to display their GUI on the visible desktop by 
requesting the resource "display_win_gui" (qsub \-l display_win_gui=1). See
.M submit 1 ,
option "\-v", for information on how to set a job environment variable.
.br
When running GUI jobs, the environment variable SGE_BACKGND_MODE can be used to
direct where the output of the GUI will be displayed. If the environment variable
SGE_BACKGND_MODE is set to "TRUE" or "1" and "\-l display_win_gui=true" is requested,
the job is started by the helper service, but the GUI is not opened on the execution 
host's screen. Instead it is opened on a virtual screen.
.PP
The Windows administrator can operate this service using the usual Windows
Service Control Manager or the "net" command. The "start" method starts the
service if possible, the "stop" method stops it regardless of any jobs that
might be still running. For a graceful shutdown, first run the "pause" method.
This will start the shutdown procedure of the service and prevent it from
accepting any further jobs. After this, the "resume" (aka "continue") method
will fail as long as there are jobs running in the
.IR SGE_Helper_Service.exe .
If no job is left in the service, the "resume" method will succeed and the
service will stop.
.br
.\"
.\"
.SH OPTIONS
.\"
.SS "\fB\-help\fP"
Prints a listing of all options.
.SS "\fB\-install\fP"
Installs the service. Only the local Administrator has the permissions to install the service.
.SS "\fB\-uninstall\fP"
Uninstalls the service. Only the local Administrator has the permissions to uninstall the service. Only a stopped service can be uninstalled. See the Windows command net.exe for information on how to stop a service.
.fi
.\"
.\"
.SH "SEE ALSO"
.M xxqs_name_sxx_intro 1 ,
.M xxqs_name_sxx_execd 8 .
.\"
.SH "COPYRIGHT"
See
.M xxqs_name_sxx_intro 1
for a full statement of rights and permissions.


NEWS
Notable changes in Son of Grid Engine
=====================================

For earlier changes (comprising Sun Grid Engine), see Changelog in the
source distribution, and the bug lists at
http://arc.liv.ac.uk/SGE/howto/Installation,%20Upgrade,%20Patches.
For detailed recent change information, including credits for changes
not listed here, see https://arc.liv.ac.uk/trac/SGE/log/sge/

"[#<n>]" below refers to ticket <n> at
https://arc.liv.ac.uk/trac/SGE/ticket/<n>.  For historical reasons,
changes due to Univa/Sun are tagged as "[U]", or "[(U)]" if partially
due to them or appearing first in this version.


Version 8.1.9
-------------

* Security
  * Drop setgid privileges correctly (CERT POS36-C); daemons now run
    with (only) admin user's groups, not root's

* Building

  * MUNGE authentication can be built in (aimk -munge)
  * Use a minimal autoconf framework
  * Fix building on recent Darwin and Cygwin [#1516, #1536, #1557]
  * Drop support for Darwin < 10 and Darwin/PPC
  * Cope with GNU/Linux ppc64le as well as ppc64el
  * Fix rpm packaging of spool objects [#1519]
  * Actually install "work" in utilbin (c.f. 8.1.8)
  * Rpm spec fixes for Fedora 23, ppc64, and OpenSuSE
  * Fix build when bool is defined in headers

* Bug fixes

  * Man fixes
  * Message fixes [#1538, #1539]
  * Fix RPM packaging of spool files [#1519]
  * Don't set port and cluster in check_sge.py, and accept memory in T
  * Fix pam_sge-qrsh-setup for recent PAM
  * Do a full, not a partial match on GECOS data when selecting certificate
    in sge_ca [#1517]
  * Fix checking of admin host in qconf [#1547]
  * Report the actual exit status of master task in tightly integrated
    parallel jobs
  * Fix share tree accounting with array jobs [#435]
  * Fix doubled multipliers of units in qacct memory reports
  * Improve job scheduling within a sharetree node [#1550]
  * Flush spool at qmaster exit [#1551]
  * Sort jobs and advance reservations when reading from spool [#1552, #1553]
  * Fix detection of old v. new type of cpuset [#1523]
  * Fix treatment of some invalid entries in variable lists or the
    environment on job submission

* Enhancements

  * MUNGE authentication (as above)
  * qconf -clearusage can act on a specific user
  * Allow generating load sensor data just before sending them
  * Add sge_ca "-revoke <cert>" option
  * Allow running DRMAA jobs as arbitrary user [#492].  (Partially
    done in v 8.1.0, but not mentioned; now works with CSP.)
  * Report SSL error numbers in hex [#1539]
  * Use PVM_VMID environment variable [#44]
  * rc templates support a "status" argument (for HA usage)

* Potentially incompatible changes

  * Print error messages to stderr and don't print on success with -terse
    to be consistent with submit(1) saying that error messages go to
    stderr and to make -terse more useful in cases like
      cat test.o$(qsub -sync y -terse -cwd -N test ...)
  * sge_ca -renew and friends no longer revoke old certificates
  * The qconf -sep and -cb options have been removed finally
  * A POSIX 2001-compatible library is required

Version 8.1.8
-------------

* Building

  * Fix build with recent GNU binutils
  * Fix rpm build on Fedora and maybe use packaged swing-layout
  * Fix dpkg build on recent Debian/Ubuntu
  * uid_t/gid_t are unsigned on NetBSD
  * Install "work" in utilbin
  * Move qsched back to main package

* Bug fixes

  * Man fixes
  * Fix buffer overflow in qmake [#1507]
  * Avoid possible link-time error (Debian bug # 749413)
  * Fix some global messages from qstat -j
  * Avoid truncating usage values in accounting records
  * Fix execd remembering existing core binding when restarted [#1511]

* Enhancements

  * Nagios and Ganglia monitoring scripts
  * Python JSV implementation
  * Upstart and systemd startup files
  * Define DRMAA_LIBRARY_PATH in setup scripts

Version 8.1.7
-------------

* Building

  * Remove the shipped old jemalloc in favour of an external one
    (aimk -jemalloc)
  * Always use system libraries -- never load SSL stuff dynamically
  * aimk no longer uses specific paths for openssl etc.  Use
    INPUT_*_FLAGS if necessary
  * SUpport building rpm on OpenSuSE and RHEL7

* Bug fixes

  * Man, other doc, and message fixes/changes [#1504]
  * Fix minor memory leaks and check result of all malloc calls
  * Don't pause for a second with client JSVs [#1475]
  * Allow longer "failed" strings in admin mail
  * Fix spurious thread error reports from qping [#588]
  * Avoid possible crash reading malformed configuration files
  * Fix GUI installer startup [#1488]
  * Always define SGE_STARTER_USE_LOGIN_SHELL with starter method
  * Modify cpuset-based process killing [#1483, #1490]
  * Make various files optional when setting permissions at installation
  * Fix cpu-bound shepherd with qlogin
  * Fix queue list for adding queue in qmon [#1495]
  * Clean up some file permissions and shebangs in non-executables [#1493]
  * Really ignore -e, -o args of qrsh/qsh/qlogin, as documented
  * Fix use of SGE_QMASTER_PIDFILE and double writing of pid file
  * Fix occasional qmaster major space leak with schedd_job_info [#360, #682]
  * Wait until daemon goes down when stopping sge_execd

* Enhancements

  * Support Cygwin x86_64, and GNU/Linux on arm64 and ppc64el (untested)
  * qconf -aattr et al now support modifying types user, userset, calendar,
    and project
  * Translate exit code to signal name in qacct output
  * qacct -m prints only master tasks
  * Add ar_sub_time to qacct output
  * Allow rctemplate configuration from /etc/sysconfig as well as
    /etc/default
  * Update MPI PE templates and install a tight integration "mpi" PE
  * Make suid files, dtrace, qmon, start_gui_installer optional in
    setfileperm.sh (used at installation)
  * Speed up JSV in bash by not calling cut

* Potentially incompatible changes

  * Default ENABLE_ADDGRP_KILL to on
  * Clean symbol exports from libdrmaa.so (GNU ld only); now only exports

source/utilbin/windows/SGE_Helper_Service/SGE_Helper_Service.cpp
/*___INFO__MARK_BEGIN__*/
/*************************************************************************
 *
 *  The Contents of this file are made available subject to the terms of
 *  the Sun Industry Standards Source License Version 1.2
 *
 *  Sun Microsystems Inc., March, 2001
 *
 *
 *  Sun Industry Standards Source License Version 1.2
 *  =================================================
 *  The contents of this file are subject to the Sun Industry Standards
 *  Source License Version 1.2 (the "License"); You may not use this file
 *  except in compliance with the License. You may obtain a copy of the
 *  License at http://gridengine.sunsource.net/Gridengine_SISSL_license.html
 *
 *  Software provided under this License is provided on an "AS IS" basis,
 *  WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING,
 *  WITHOUT LIMITATION, WARRANTIES THAT THE SOFTWARE IS FREE OF DEFECTS,
 *  MERCHANTABLE, FIT FOR A PARTICULAR PURPOSE, OR NON-INFRINGING.
 *  See the License for the specific provisions governing your rights and
 *  obligations concerning the Software.
 *
 *   The Initial Developer of the Original Code is: Sun Microsystems, Inc.
 *
 *   Copyright: 2001 by Sun Microsystems, Inc.
 *
 *   All Rights Reserved.
 *
 * Portions of this software are Copyright (c) 2011 Univa Corporation
 *
 ************************************************************************/
/*___INFO__MARK_END__*/

#include <winsock2.h>
#include <stdio.h>

#include "Job.h"
#include "JobList.h"
#include "JobStart.h"
#include "SGE_Helper_Service.h"
#include "ServiceInstaller.h"
#include "Communication.h"
#include "Logging.h"

// constants
// name, display name and description of the service in "Services" dialog
// of control panel
static const char g_szServiceName[] = "SGE_Helper_Service.exe";
static const char g_szServiceDisplayName[] = "Sun Grid Engine Helper Service";
static const char g_szServiceDescription[] = "Enables Sun Grid Engine "
                               "execution host functionality to let jobs "
                               "display their GUI to the visible desktop.";

// internal name of the service handler
static const char g_szServiceHandlerName[] = "Sun Grid Engine Helper Service Handler";

// global variables
HANDLE          g_hWinstaACLMutex = NULL;
HANDLE          g_hDeskACLMutex = NULL;
C_Communication g_Communication;
C_JobList       g_JobList;
BOOL            g_bAcceptJobs    = TRUE;
int             g_Port = 0;
BOOL            g_bDoLogging = FALSE;

// module variables
static SERVICE_STATUS_HANDLE g_hServiceStatus = NULL;
static BOOL                  g_bRunning       = TRUE;

// function forward declarations
static void  WINAPI SunGridEngineHelperServiceStart(DWORD dwArgc, LPTSTR *lpszArgv);
static DWORD WINAPI SunGridEngineHelperServiceHandlerEx(DWORD dwControl, 
                          DWORD dwEventType, LPVOID lpEventData, LPVOID lpContext);
static void PrintHelp(const char *pszExecutableName, const char *pszServiceName);
static void ServiceError();


/****** main() ****************************************************************
*  NAME
*     main() -- starting point of the executable
*
*  SYNOPSIS
*     int main(int argc, char *argv[]);
*
*  FUNCTION
*     A Windows Service works this way:
*     When the service is started, either automatically at boot time or manually,
*     main() gets called by the ServiceControlManager (SCM).
*
*     main() has to register the Control Dispatcher Thread to the SCM. The Control
*     Dispatcher Thread gets called by the SCM, 
*
*  INPUTS
*     int  argc    - number of command line arguments
*     char *argv[] - array of arguments
*
*  RESULT
*     int - exit status of the executable
*     0: no errors
*     1: installation or uninstallation of service failed
* TODO: Document all error codes
*
*  NOTES
*******************************************************************************/
int main(int argc, char *argv[])
{
   int ret = 0;

   SERVICE_TABLE_ENTRY DispatchTable[] = {
      {(LPSTR)g_szServiceName, SunGridEngineHelperServiceStart},
      {NULL,                   NULL}};

   if(argc==2) {
      /*
       * Newer Microsoft compilers warn if "strnicmp" is used, you have to use "_strnicmp" to
       * avoid this warning. This is for all libc functions.
       */
      if(_strnicmp(argv[1], "-install", 8)==0) {
         ret = InstallService(argv[0], g_szServiceName,
                              g_szServiceDisplayName, g_szServiceDescription);
      } else if(_strnicmp(argv[1], "-uninstall", 10)==0) {
         ret = UninstallService(g_szServiceName);
      } else if((g_Port=atoi(argv[1]))!=0) {
         if(!StartServiceCtrlDispatcher(DispatchTable)) {
            SvcDebugOut("[%s] StartServiceCtrlDispatcher error = %d\n",
               g_szServiceDisplayName, GetLastError());
            ret = 7;
         }
      } else {
         PrintHelp(argv[0], g_szServiceDisplayName);
      }
   } else if(argc==1) {
      // Start Execd Helper Service
      if(!StartServiceCtrlDispatcher(DispatchTable)) {
         SvcDebugOut("[%s] StartServiceCtrlDispatcher error = %d\n",
            g_szServiceDisplayName, GetLastError());
         ret = 7;
      }
   } else {
      PrintHelp(argv[0], g_szServiceDisplayName);
   }
	return ret;
}

/****** SunGridEngineHelperServiceHandlerEx() *********************************
*  NAME
*     SunGridEngineHelperServiceHandlerEx() -- handler for the service that is
*                                      invoked by the service control manager
*
*  SYNOPSIS
*     DWORD WINAPI SunGridEngineHelperServiceHandlerEx(DWORD  dwControl,
*                                                     DWORD  dwEventType,
*                                                     LPVOID lpEventData,
*                                                     LPVOID lpContext)
*
*  FUNCTION
*     This handler is invoked by the service control manager when the service
*     is to be started, stopped or altered.
*     See HandlerEx() in Microsoft PlatformSDK documentation.
*  
*     Pause and continue are used to shutdown the service gracefully:
*     First call pause, the service goes to pause mode and doesn't accept any
*     new jobs. After this, call resume. Resume will fail as long as there
*     are jobs running - if no job is left running, resume will succeed and
*     stop the service.
*
*  INPUTS
*     DWORD  dwControl   - action that is to be performed
*     DWORD  dwEventType - more specific information for some actions
*     LPVOID lpEventData - more specific information for some actions
*     LPVOID lpContext   - more specific information for some actions
*
*  RESULT
*     DWORD - NO_ERROR if no errors occured, see HandlerEx() in Microsoft
*             PlatformSDK documentation for list of errors.
*
*  NOTES
*******************************************************************************/
static DWORD WINAPI SunGridEngineHelperServiceHandlerEx(DWORD  dwControl,
                                                       DWORD  dwEventType,
                                                       LPVOID lpEventData,
                                                       LPVOID lpContext)
{
   SERVICE_STATUS  SvcStatus;

   SvcStatus.dwServiceType             = SERVICE_WIN32_OWN_PROCESS 
                                         |SERVICE_INTERACTIVE_PROCESS;
   SvcStatus.dwControlsAccepted        = SERVICE_ACCEPT_SHUTDOWN  
                                         |SERVICE_ACCEPT_STOP 
                                         |SERVICE_ACCEPT_PAUSE_CONTINUE;
   SvcStatus.dwWin32ExitCode           = NO_ERROR;
   SvcStatus.dwServiceSpecificExitCode = 0;
   SvcStatus.dwCheckPoint              = 0;
   SvcStatus.dwWaitHint                = 1000;

   switch(dwControl) {
      case SERVICE_CONTROL_STOP:
         SvcStatus.dwCurrentState = SERVICE_STOPPED; 

         if(!SetServiceStatus(g_hServiceStatus, &SvcStatus)) { 
            ServiceError();
         } 
         g_bRunning = FALSE;
         break;

      case SERVICE_CONTROL_PAUSE:
         g_bAcceptJobs = FALSE;
         SvcStatus.dwCurrentState = SERVICE_PAUSED; 

         if(!SetServiceStatus(g_hServiceStatus, &SvcStatus)) { 
            ServiceError();
         } 
         break;

      case SERVICE_CONTROL_CONTINUE:
         if(g_JobList.IsEmpty()) {
            g_bRunning = FALSE;
            SvcStatus.dwCurrentState = SERVICE_STOPPED;
         } else {
            SvcStatus.dwCurrentState = SERVICE_PAUSED; 
         }

         if(!SetServiceStatus(g_hServiceStatus, &SvcStatus)) { 
            ServiceError();
         } 
         break;

      case SERVICE_CONTROL_INTERROGATE:
         break; 

      default: 
         SvcDebugOut("[%s] Unrecognized opcode %ld\n",  
                       g_szServiceDisplayName, dwControl); 
   }
   return NO_ERROR;
}

/****** SunGridEngineHelperServiceStart() *********************************************
*  NAME
*     SunGridEngineHelperServiceStart() -- starting point of the service
*
*  SYNOPSIS
*     void WINAPI SunGridEngineHelperServiceStart(DWORD dwArgc, LPTSTR *lpszArgv)
*
*  FUNCTION
*     This is the starting point of the service that is called by the
*     service manager when the service is started.
*
*  INPUTS
*     DWORD  dwArgc    - number of arguments
*     LPTSTR *lpszArgv - argument list
*
*  RESULT
*     void - no result
*
*  NOTES
*******************************************************************************/
static void WINAPI SunGridEngineHelperServiceStart(DWORD dwArgc, LPTSTR *lpszArgv)
{
   SERVICE_STATUS        ServiceStatus;
   DWORD                 dwStatus = 0;
   int                   ret;
   char                  job_result[] = "done";

   // Parse command line options
   for (DWORD i=1; i<dwArgc; i++) {
      if (strcmp(lpszArgv[i], "-log") == 0) {
         g_bDoLogging = TRUE;
         break;
      }
   }
   WriteToLogFile("Starting %s.", g_szServiceDisplayName);

   ServiceStatus.dwServiceType             = SERVICE_WIN32_OWN_PROCESS
                                             |SERVICE_INTERACTIVE_PROCESS;
   ServiceStatus.dwCurrentState            = SERVICE_RUNNING;
   ServiceStatus.dwControlsAccepted        = SERVICE_ACCEPT_SHUTDOWN
                                             |SERVICE_ACCEPT_STOP  
                                             |SERVICE_ACCEPT_PAUSE_CONTINUE;
   ServiceStatus.dwWin32ExitCode           = NO_ERROR;
   ServiceStatus.dwServiceSpecificExitCode = 0;
   ServiceStatus.dwCheckPoint              = 0;
   ServiceStatus.dwWaitHint                = 1000;

   g_hServiceStatus = RegisterServiceCtrlHandlerEx(g_szServiceName, 
                              SunGridEngineHelperServiceHandlerEx, NULL);

   if(g_hServiceStatus == (SERVICE_STATUS_HANDLE)0) { 
      SvcDebugOut("[%s] RegisterServiceCtrlHandler failed %d\n",  
                             g_szServiceDisplayName, GetLastError()); 
      return; 
   } 

   if(!SetServiceStatus(g_hServiceStatus, &ServiceStatus)) {
      dwStatus = GetLastError(); 
      SvcDebugOut("[%s] SetServiceStatus error %ld\n", 
                         g_szServiceDisplayName, dwStatus); 
   }

   WriteToLogFile("%s started successfully.", g_szServiceDisplayName);

   g_hWinstaACLMutex = CreateMutex(NULL, FALSE, "WinstaACLMutex");
   g_hDeskACLMutex   = CreateMutex(NULL, FALSE, "DeskACLMutex");

   // listen on port
   ret = g_Communication.InitComm();
   if(ret != 0) {
      g_bRunning = FALSE;
   }

   // Main loop
   while(g_bRunning == TRUE) {
      ret = g_Communication.DoComm();
      if(ret != 0) {
         g_bRunning = FALSE;
      }
   }
   g_Communication.ExitComm();

   CloseHandle(g_hWinstaACLMutex);
   CloseHandle(g_hDeskACLMutex);
}

/****** PrintHelp() ***********************************************************
*  NAME
*     PrintHelp() - prints usage information to stdout
*
*  SYNOPSIS
*     void PrintHelp(const char *pszExecutableName, const char *pszServiceName)
*
*  FUNCTION
*     Prints usage information to stdout
*
*  INPUTS
*     const char* pszExecutableName - name of the executable (use argv[0])
*     const char* pszServiceName    - display name of the service
*
*  RESULT
*     void - no result
*
*  NOTES
*******************************************************************************/
static void PrintHelp(const char *pszExecutableName, const char *pszServiceName)
{
   printf("%s -install | -uninstall | -help\n"
          "  -install     Installs %s\n"
          "  -uninstall   Uninstalls %s\n"
          "  -help        Displays this help\n",
          pszExecutableName, pszServiceName, pszServiceName);
}

/****** SvcDebugOut() *********************************************************
*  NAME
*     SvcDebugOutput() - prints error messages to the debugger
*
*  SYNOPSIS
*     void SvcDebugOut(LPSTR String, ...)
*
*  FUNCTION
*     Prints error messages to the debugger
*
*  INPUTS
*     LPSTR  String      - format string of error message
*     ...                - parameters of error message
*
*  RESULT
*     void - no result
*
*  NOTES
*******************************************************************************/
void SvcDebugOut(LPSTR String, ...)
{
   char    Buffer[1024];
   va_list args;

   va_start(args, String);
   _vsnprintf(Buffer, 1023, String, args);

   OutputDebugStringA(Buffer);
}

/****** ServiceError() ********************************************************
*  NAME
*     ServiceError() -- Writes the last error to the debugger
*
*  SYNOPSIS
*     static void ServiceError()
*
*  FUNCTION
*     Prints the last error (GetLastError()) to the debugger.
*
*  RESULT
*     void - no result
*******************************************************************************/
static void ServiceError()
{
   DWORD dwStatus = GetLastError(); 
   SvcDebugOut("[%s] SetServiceStatus error %ld\n", 
      g_szServiceDisplayName, dwStatus); 
}
